pkg/controller/knativeserving/common/images.go

```diff
@@ -0,0 +1,57 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package common
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/go-logr/logr"
+	servingv1alpha1 "github.com/knative/serving-operator/pkg/apis/serving/v1alpha1"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// UpdateImage updates the image of the deployment with a new registry and tag
+// @param image takes the form of new-registry.io/full/path/${NAME}:new-tag
+func UpdateImage(deployment *appsv1.Deployment, registry *servingv1alpha1.Registry, log logr.Logger) {
+	image := imageName(deployment.GetName(), registry)
+	if image == "" {
+		return
+	}
+	image = strings.ReplaceAll(image, "${NAME}", deployment.GetName())
+	containers := deployment.Spec.Template.Spec.Containers
+	for index := range containers {
+		container := &containers[index]
+		updateContainer(container, image, log)
```

Comment:
This formulation means that any Pod that has more than one container will run the same image for all containers. That's probably not what is desired -- maybe you want to use the name of the container, rather than the name of the Deployment?

Reply:
You bring up a good point that this string will override all containers. We have a few options. We can either put the ${NAME} interpolation back into the override which will use the container name instead of the deployment name. Or we can have individual overrides per container. I think the interpolation is easier to understand and consume, however it makes it more confusing or unusable if the images are pretty unrelated and now need to follow the same registry and path.

Reply:
Personally, I think the keys in the override map should be container names rather than deployment names, since only containers have images.
That being said, these transforms will only apply to the upstream manifest packaged in kodata. And as of 0.7.0, all deployments have a single container and their names are identical. If we expect that to change in a subsequent release, we should absolutely account for it now. But if we don't, I'm not sure adding the complexity now is worth it.

Reply:
Do we expect deployments A and B to each have a container C anytime soon? Or is it sufficient to expect a naming convention of A-C and B-C for those containers? Personally, I think the Default is going to account for 99% of the usecases. I suppose we could consider the Override to be a map of maps, with the first key being the Deployment name and the second being the Container name?

Reply:
I agree. I think the map of maps is likely excessive right now. It would add a lot of additional complexity and current duplication:

```yaml
overrides:
  activator:
    activator: "new-activator-image"
  controller:
    controller: "new-controller-image"
```

However if one of the containers required a security fix in a deployment with multiple containers, the tags will need to be updated on both those containers which is non-obvious. This is especially concerning as the user would need to know this while addressing a potential production issue.

Reply:
But as long as all containers are uniquely named within the entire manifest, it's not an issue, correct?
And again, you guys are worrying me by suggesting we might have a deployment with multiple containers? Why do you expect that to happen? We control the manifest, not the user.

Reply:
I do like the idea of using container names rather than deployment names here - should be relatively simple to implement and somewhat future proofs us if we add additional containers to a deployment in serving sometime. We probably want a static whitelist as well though of containers we expect to see. I dont think we need to worry too much about the nuances of what happens when serving runs multiple containers and someone wants to roll out a security fix etc - we're targeting a known quantity at this point and can worry about that if/when serving creates that case.

Reply:
I think using container name here should be fine, assuming that we communicate to the serving team that expectation. I worry that not designing for Pods with more than one container will probably result in a spectacular failure at some future point, but I think enforcing that container names are a unique mapping with images is probably reasonable.

**Specific scenario I'm thinking about:**
Imagine there is some sort of "helper" container (e.g. fluentd for copying logs) that we want to run alongside several components. I think using the same container name might be convenient here to allow updating the helper across several different deployments.

Reply:
Updated the code to use container names instead of the deployment name.

Reply:
> **Specific scenario I'm thinking about:**
> Imagine there is some sort of "helper" container (e.g. fluentd for copying logs) that we want to run alongside several components. I think using the same container name might be convenient here to allow updating the helper across several different deployments.

In that scenario, I would expect all of those fluentd containers across all those deployments to use the same image, so letting them all have the same name should be fine, I think. It's only the like-named containers that serve different purposes where i think the container-keyed override design might break down, but I'm not convinced that's a very likely scenario.